server/src/main/java/com/cloud/storage/StorageManagerImpl.java

```diff
@@ -1633,6 +1640,223 @@ public PrimaryDataStoreInfo cancelPrimaryStorageForMaintenance(CancelPrimaryStor
+                }
+            } else {
+                dataStoreVO = createChildDatastoreVO(datastoreClusterPool, childDataStoreAnswer);
+            }
+            updateStoragePoolHostVOAndBytes(dataStoreVO, hostId, childDataStoreAnswer);
+        }
+
+        handleRemoveChildStoragePoolFromDatastoreCluster(childDatastoreUUIDs);
+    }
+
+    private StoragePoolVO createChildDatastoreVO(StoragePoolVO datastoreClusterPool, ModifyStoragePoolAnswer childDataStoreAnswer) {
+        StoragePoolInfo childStoragePoolInfo = childDataStoreAnswer.getPoolInfo();
+        List<String> storageTags = _storagePoolTagsDao.getStoragePoolTags(datastoreClusterPool.getId());
+
+        StoragePoolVO dataStoreVO = new StoragePoolVO();
+        dataStoreVO.setStorageProviderName(datastoreClusterPool.getStorageProviderName());
+        dataStoreVO.setHostAddress(childStoragePoolInfo.getHost());
+        dataStoreVO.setPoolType(Storage.StoragePoolType.PreSetup);
+        dataStoreVO.setPath(childStoragePoolInfo.getHostPath());
+        dataStoreVO.setPort(datastoreClusterPool.getPort());
+        dataStoreVO.setName(childStoragePoolInfo.getName());
+        dataStoreVO.setUuid(childStoragePoolInfo.getUuid());
+        dataStoreVO.setDataCenterId(datastoreClusterPool.getDataCenterId());
+        dataStoreVO.setPodId(datastoreClusterPool.getPodId());
+        dataStoreVO.setClusterId(datastoreClusterPool.getClusterId());
+        dataStoreVO.setStatus(StoragePoolStatus.Up);
+        dataStoreVO.setUserInfo(datastoreClusterPool.getUserInfo());
+        dataStoreVO.setManaged(datastoreClusterPool.isManaged());
+        dataStoreVO.setCapacityIops(datastoreClusterPool.getCapacityIops());
+        dataStoreVO.setCapacityBytes(childDataStoreAnswer.getPoolInfo().getCapacityBytes());
+        dataStoreVO.setUsedBytes(childDataStoreAnswer.getPoolInfo().getCapacityBytes() - childDataStoreAnswer.getPoolInfo().getAvailableBytes());
+        dataStoreVO.setHypervisor(datastoreClusterPool.getHypervisor());
+        dataStoreVO.setScope(datastoreClusterPool.getScope());
+        dataStoreVO.setParent(datastoreClusterPool.getId());
+
+        Map<String, String> details = new HashMap<>();
+        if(org.apache.commons.lang.StringUtils.isNotEmpty(childDataStoreAnswer.getPoolType())) {
+            details.put("pool_type", childDataStoreAnswer.getPoolType());
+        }
+        _storagePoolDao.persist(dataStoreVO, details, storageTags);
+        return dataStoreVO;
+    }
+
+    private void handleRemoveChildStoragePoolFromDatastoreCluster(Set<String> childDatastoreUUIDs) {
+
+        for (String childDatastoreUUID : childDatastoreUUIDs) {
+            StoragePoolVO dataStoreVO = _storagePoolDao.findPoolByUUID(childDatastoreUUID);
+            List<VolumeVO> allVolumes = _volumeDao.findByPoolId(dataStoreVO.getId());
+
+            for (VolumeVO volume : allVolumes) {
+                VMInstanceVO vmInstance = _vmInstanceDao.findById(volume.getInstanceId());
```

Comment:
_volume.getInstanceId()_ is null for detached disks

Reply:
Filtered the volumes list by removing volumes which does not have instance id and keeping only volumes which are in ready state.